components/forwarder/forwarder.go

```diff
@@ -0,0 +1,127 @@
+	publisher message.Publisher
+	logger    watermill.LoggerAdapter
+	config    Config
+}
+
+// NewForwarder creates a forwarder which will subscribe to the topic provided in the config using the provided subscriber.
+// It will publish messages received on this subscription to the destination topic embedded in the enveloped message using the provided publisher.
+//
+// Provided subscriber and publisher can be from different Watermill Pub/Sub implementations, i.e. MySQL subscriber and Google Pub/Sub publisher.
+//
+// Note: Keep in mind that by default the forwarder will nack all messages which weren't sent using a decorated publisher.
+// You can change this behavior by passing a middleware which will ack them instead.
+func NewForwarder(subscriberIn message.Subscriber, publisherOut message.Publisher, logger watermill.LoggerAdapter, config Config) (*Forwarder, error) {
+	config.setDefaults()
+
+	routerConfig := message.RouterConfig{CloseTimeout: config.CloseTimeout}
+	if err := routerConfig.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid router config")
+	}
+
+	router, err := message.NewRouter(routerConfig, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create a router")
+	}
+
+	f := &Forwarder{router, publisherOut, logger, config}
+
+	router.AddNoPublisherHandler(
+		"events_forwarder",
+		config.ForwarderTopic,
+		subscriberIn,
+		f.forwardMessage,
+	)
+
+	router.AddMiddleware(config.Middlewares...)
+
+	return f, nil
+}
+
+// Run runs forwarder's handler responsible for forwarding messages.
+// This call is blocking while the forwarder is running.
+// ctx will be propagated to the forwarder's subscription.
+//
+// To stop Run() you should call Close() on the forwarder.
+func (f *Forwarder) Run(ctx context.Context) error {
+	return f.router.Run(ctx)
+}
+
+// Close stops forwarder's handler.
+func (f *Forwarder) Close() error {
+	return f.router.Close()
+}
+
+// Running returns channel which is closed when the forwarder is running.
+func (f *Forwarder) Running() chan struct{} {
+	return f.router.Running()
+}
+
+// DecoratePublisher is a helper method which decorates a given publisher so when it publishes messages,
+// it envelopes them and publish directly to the forwarder's topic provided in the forwarder's config.
+//
+// Enveloping means that the message is put into the generic envelope containing the original message
+// and a destination topic taken from publisher `Publish` method. The destination topic is used later
+// by the forwarder to forward it to a specific topic.
+func (f *Forwarder) DecoratePublisher(publisherIn message.Publisher) message.Publisher {
+	return NewPublisherDecorator(publisherIn, PublisherConfig{
+		ForwarderTopic: f.config.ForwarderTopic,
+	})
+}
+
+func (f *Forwarder) forwardMessage(msg *message.Message) error {
+	destTopic, unwrappedMsg, err := unwrapMessageFromEnvelope(msg)
+	if err != nil {
+		return errors.Wrap(err, "cannot unwrap message from an envelope")
```

Comment:
if you are returning error it will nack 4ever ;) I think that it may be a good idea to inject Watermill logger here and log error

Reply:
and log the payload! so in case if this msg should be recovered it will not be lost

Reply:
> Note: Keep in mind that by default the forwarder will nack all messages which weren't sent using a decorated publisher.

ah, like that, so maybe it will be a good idea to have boolean in config to encapsulate adding middleware?

Reply:
> boolean in config to encapsulate adding middleware 
Sounds like a good idea :)

Reply:
71f6a78